lib/kennel/parts_writer.rb

```diff
@@ -0,0 +1,70 @@
+# frozen_string_literal: true
+
+require "find"
+
+module Kennel
+  class PartsWriter
+    def initialize(filter:)
+      @filter = filter
+    end
+
+    def write(parts)
+      Progress.progress "Storing" do
+        if filter.tracking_id_filter
+          write_changed(parts)
+        else
+          old = old_paths
+          used = write_changed(parts)
+          (old - used).each { |p| FileUtils.rm_rf(p) }
+        end
+      end
+    end
+
+    private
+
+    attr_reader :filter
+
+    def write_changed(parts)
+      used = ["generated"]
+
+      Utils.parallel(parts, max: 2) do |part|
+        path = "generated/#{part.tracking_id.tr("/", ":").sub(":", "/")}.json"
+        used.concat [File.dirname(path), path] # only 1 level of sub folders, so this is safe
+        payload = part.as_json.merge(api_resource: part.class.api_resource)
+        write_file_if_necessary(path, JSON.pretty_generate(payload) << "\n")
+      end
+
+      used
+    end
+
+    def apply_cleanup_to
+      if filter.project_filter
+        filter.project_filter.map { |project| "generated/#{project}" }
+      else
+        ["generated"]
+      end
+    end
+
+    def old_paths
+      apply_cleanup_to.flat_map do |path|
+        if File.exist?(path)
+          Find.find(path).to_a
+        else
+          []
+        end
+      end
+    end
```

Comment:
finding all files adds overhead to the "generate all" usecase, so prefer to avoid

Reply:
```
$ time ruby -rfind -rbenchmark -e 'puts Benchmark.measure { 100.times { Find.find("generated").to_a } }'
  5.003687   8.399515  13.403202 ( 13.436303)
ruby -rfind -rbenchmark -e   5,07s user 8,43s system 99% cpu 13,541 total
```

It seems to be inconsequential. Premature optimization and all that.

Reply:
so 100ms right ? ... prefer to avoid that
... also means lots of disk-io
I'm hoping that there is a cheaper way of doing this ... ideally in it's own PR so we can code-golf more
... if it's complicated to take back out then just leave it and I'll take a look once it's merged

Reply:
Given that the pre-PR version is buggy, let's concentrate on correctness first, and maybe optimize later.